src/append_only_data.rs

```diff
@@ -206,10 +206,39 @@ impl Permissions for UnpubPermissions {
         self.owner_entry_index
     }
 
-    fn check_permission(&self, requester: PublicKey, action: Action) -> bool {
-        match self.permissions.get(&requester) {
-            Some(perm) => perm.is_allowed(action),
-            None => false,
+    fn check_permissions_for_key(&self, requester: PublicKey, request: &Request) -> Result<()> {
+        match request {
+            Request::GetAData(..)
+            | Request::GetADataShell { .. }
+            | Request::GetADataRange { .. }
+            | Request::GetADataIndices(..)
+            | Request::GetADataLastEntry(..)
+            | Request::GetADataPermissions { .. }
+            | Request::GetUnpubADataUserPermissions { .. }
+            | Request::GetADataOwners { .. } => {
+                if self.is_action_allowed(requester, Action::Read) {
```

Comment:
Seems like it might be simpler if we changed `is_action_allowed()` to return a `Result<()>` too?